Remove reactn dependency from @labkey/components package

cnathe opened the pull request:
#### Rationale
With the completion of the migration to QueryModel, the only two things left in reactn global state were Notifications and a BrowserHistory object. This PR updates those two cases so that we no longer use the reactn package for global state. Notifications have been converted to a React Context provider. And BrowserHistory has been converted to a global variable in util.global.ts.

#### Related Pull Requests
* https://github.com/LabKey/labkey-ui-components/pull/898
* https://github.com/LabKey/labbook/pull/228
* https://github.com/LabKey/platform/pull/3525
* https://github.com/LabKey/inventory/pull/482
* https://github.com/LabKey/biologics/pull/1438
* https://github.com/LabKey/sampleManagement/pull/1095

#### Changes
* convert notifications to use a NotificationsContext (plus useNotificationsContext and withNotificationsContext)
* EditableGridPanelForUpdate update so that it stores and display data errors as an Alert instead of having app use createNotification

cnathe committed:
convert notifications to use a NotificationsContext (plus useNotificationsContext and withNotificationsContext)

cnathe committed:
v2.195.1-fb-removeReactnDep.0

cnathe committed:
jest fixes round 1: more use of mountWithAppServerContext and removal of initNotificationsState

cnathe committed:
jest fixes round 2: more use of mountWithAppServerContext

cnathe committed:
jest fixes round 3: allow for initial notifications in context

cnathe committed:
v2.195.1-fb-removeReactnDep.1

cnathe committed:
export NotificationsContextProvider

cnathe committed:
v2.195.1-fb-removeReactnDep.2

cnathe committed:
fix for selenium test: dismissNotifications was outdated when called from move storage unit sequence

cnathe committed:
v2.195.1-fb-removeReactnDep.3

cnathe committed:
revert previous change and convert dismissNotifications to work with current state instead of as dependency

cnathe committed:
v2.195.1-fb-removeReactnDep.4

cnathe committed:
Merge branch 'develop' into fb_removeReactnDep

# Conflicts:
#	packages/components/package.json
#	packages/components/releaseNotes/components.md

cnathe committed:
v2.196.0-fb-removeReactnDep.0

cnathe committed:
Fix for LKS field editor usage of NameExpressionGenIdBanner which does not wrap notifications context provider

cnathe committed:
v2.196.0-fb-removeReactnDep.1

cnathe committed:
EditableGridPanelForUpdate update so that it stores and display data errors as an Alert instead of having app use createNotification

cnathe committed:
v2.196.0-fb-removeReactnDep.2

cnathe committed:
finish conversion of global BrowserHistory

cnathe committed:
v2.196.0-fb-removeReactnDep.3

cnathe committed:
update error message

cnathe committed:
Merge branch 'develop' into fb_removeReactnDep

# Conflicts:
#	packages/components/package.json
#	packages/components/releaseNotes/components.md

cnathe committed:
v2.196.1-fb-removeReactnDep.0

cnathe committed:
fix for sample editable grid consumed selenium test failure

cnathe committed:
jest test fix

cnathe committed:
v2.196.1-fb-removeReactnDep.1

labkey-alan commented on packages/components/src/internal/app/reducers.ts:
```diff
@@ -112,6 +112,7 @@ export const ProductMenuReducers = handleActions<ProductMenuState, any>(
     new ProductMenuModel()
 );
 
+// TODO is this still used?
```
It is used in Biologics.

labkey-alan commented on packages/components/src/internal/components/administration/BasePermissions.spec.tsx:
```diff
@@ -22,10 +21,6 @@ import { BasePermissionsImpl, BasePermissionsImplProps } from './BasePermissions
 const TEST_POLICY = SecurityPolicy.create(policyJSON);
 
-beforeAll(() => {
-    initNotificationsState();
```
So glad to not have to do this anymore!

cnathe replied:
Agreed!

labkey-alan commented on packages/components/src/internal/components/administration/UserManagement.spec.tsx:
```diff
@@ -52,15 +50,15 @@ describe('UserManagement', () => {
 
     test('default props', () => {
-        const wrapper = mountWithAppServerContext(<UserManagementPageImpl {...getDefaultProps()} />, undefined, {
+        const wrapper = mountWithAppServerContext(<UserManagementPageImpl {...getDefaultProps()} createNotification={jest.fn} dismissNotifications={jest.fn} />, {}, {
```
These should be jest.fn()

cnathe replied:
done

labkey-alan commented on packages/components/src/internal/components/assay/AssayResultDeleteModal.spec.tsx:
```diff
@@ -21,5 +21,7 @@ describe('<AssayResultDeleteModal/>', () => {
         // Progress
         expect(wrapper.childAt(1).prop('title')).toBe('Deleting 1 assay result');
+
+        wrapper.unmount();
```
Unless you're testing that something specific is happening during component unmount (e.g. removing some sort of listener, canceling API requests) then you don't need to unmount the component. As far as I can tell we're not doing that.

cnathe replied:
got it. removed. we have unmount() calls sprinkled throughout our jest tests. I thought for awhile it was required and would cause issues with downstream tests if not done...maybe that was never the case?

labkey-susanh replied:
I have read (perhaps [here](https://github.com/enzymejs/enzyme/issues/911), perhaps out of date) that calling unmount() is actually best practice for more efficient jest test runs.

labkey-alan replied:
I'm going to investigate implementing the generic solution outlined in that issue on Monday. I would much prefer it to be something that is automatically handled for us.

labkey-alan commented on packages/components/src/internal/components/notifications/Notification.spec.tsx:
```diff
@@ -15,31 +15,78 @@
 
 beforeEach(() => {
-    notificationInit();
+    LABKEY.moduleContext = {
```
All of these `LABKEY.` usages should be replaced with constants that are passed as the server context variable to the `mountWithAppServerContext` calls below.

cnathe replied:
I tried, but that doesn't seem to work. The usage of the `LABKEY.container.formats` value that is being set in this way is in `src/internal/util/Date.ts` at `getDateFormat` line 81. It looks to be using `getServerContext` from @labkey/api.

labkey-alan replied:
We can fix it in a separate PR if you want, but the solution is relatively straightforward, we need to add an optional `context: ServerContext` argument to getDateFormat (and probably the related methods, then in `NotificationsImpl` we use `const context = useServerContext();` and pass that to `getDateFormat(context)`. We already do this for a lot of similar methods we have on App that use `moduleContext`.

cnathe replied:
Yeah, good call. That should be a simple change. Trying now.

cnathe replied:
Well...even easier. `getDateFormat` was already setup to take in an optional `container` object and `Notifications.tsx` was already using `useServerContext` so I just needed to get and pass `container` from there. Done.

labkey-alan replied:
I don't know why I didn't suggest that 😂

labkey-alan commented on packages/components/src/internal/components/notifications/Notifications.tsx:
```diff
@@ -121,11 +113,3 @@ const NotificationsImpl: FC<NotificationProps> = ({ notifications }) => {
 };
-
-// In the future when we get rid of ReactN we'll probably store Notifications in a context, so we'll remove this
-// component and add something like const { notifications } = useNotificationsContext() in NotificationsImpl above.
```
I didn't even write this comment that long ago 😄

cnathe replied:
Glad it was there. Pointed me in the right direction for sure!

labkey-alan commented on packages/components/src/internal/components/notifications/NotificationsContext.tsx:
```diff
@@ -0,0 +1,120 @@
+                    item = NotificationItemModel.create(creatable as NotificationItemProps);
+                } else item = creatable;
+
+                if (item) {
```
Because of the way we've typed `NotificationCreatable` this should always be true, so what I would suggest is dropping the `else item = creatable;` above, and changing the definition of `item` to `                let item: NotificationItemModel = createable;`. Then you can drop this if.

cnathe replied:
The type checking isn't a fan of that change. 

<img width="706" alt="Screen Shot 2022-07-13 at 3 28 04 PM" src="https://user-images.githubusercontent.com/6411206/178828597-1d7db59b-a70f-48db-a414-3b0b2eabc33d.png">

labkey-alan commented on packages/components/src/internal/components/notifications/messaging.tsx:
```diff
@@ -1,25 +1,12 @@
+import React, { ReactNode } from 'react';
 
 import { getActionErrorMessage } from '../../..';
 
```
I think everything in this file could live inside of `util/messaging.tsx`

cnathe replied:
done

labkey-alan commented on packages/components/src/internal/components/notifications/messaging.tsx:
```diff
@@ -1,25 +1,12 @@
-    count?: number,
-    additionalInfo?: string,
-) {
+export function getDeleteSuccessNotification(noun: string, count?: number, additionalInfo?: string): string {
```
I think this function, and the one below should have a different naming convention. `get` implies that we are retrieving something already created, just stored in memory, while `Notification` implies it's going to return a `Notification` model or related. I think these should be renamed to `deleteSuccessMessage`, and `deleteErrorMessage`. I think we should also consider renaming `getActionErrorMessage` to `actionErrorMessage`, however that change is a bit bigger.

cnathe replied:
Yeah, I struggled to come up with a good name for these. I like your suggestions of `deleteSuccessMessage` and `deleteErrorMessage`. I'll go with that, but likely leave `getActionErrorMessage` alone as part of this PR (I'll add a TODO).

labkey-alan commented on packages/components/src/internal/components/samples/SampleSetDeleteModal.tsx:
```diff
@@ -30,12 +31,15 @@ export const SampleSetDeleteModal: FC<Props> = props => {
             await deleteSampleSet(rowId, containerPath);
             afterDelete?.(true);
-            createDeleteSuccessNotification(' sample type');
+            createNotification(getDeleteSuccessNotification(' sample type'));
```
```suggestion
            createNotification(getDeleteSuccessNotification('sample type'));
```

cnathe replied:
done

labkey-alan commented on packages/components/src/internal/components/samples/SamplesBulkUpdateForm.spec.tsx:
```diff
@@ -96,6 +96,8 @@ describe('SamplesBulkUpdateForm', () => {
         api: getSamplesTestAPIWrapper(),
         user: TEST_USER_EDITOR,
+        createNotification: jest.fn,
+        dismissNotifications: jest.fn,
```
these should be jest.fn()

cnathe replied:
done

labkey-alan commented on packages/components/src/internal/util/global.ts:
```diff
@@ -13,34 +13,28 @@
  * limitations under the License.
  */
-import { getGlobal, setGlobal } from 'reactn';
+// import { getGlobal, setGlobal } from 'reactn';
```
can delete this :-)

cnathe replied:
hurray! done

labkey-alan commented on packages/components/src/internal/util/global.ts:
```diff
@@ -13,34 +13,28 @@
-        );
-    }
+    // add a no-op listener just to connect this global state history to the url changes
+    _BrowserHistory.listen((location, action) => {});
```
I believe we can remove this, this was needed for ReactN.

cnathe replied:
done

cnathe replied:
Turns out we do need this! After I removed it, some clicks/links in the app no logger triggered route changes. Added back in and things started working again as expected.

labkey-alan commented on packages/components/src/internal/util/global.ts:
```diff
@@ -13,34 +13,28 @@
 /**
  * Initialize the global state object for this package.
  */
 export function initBrowserHistoryState() {
```
I think we should get rid of this function and just make the line above look like:

```ts
const _BrowserHistory =  createHistory();
```

This would remove the need to ever call initQueryGridState in tests, and could remove the need to throw an error below.

cnathe replied:
I like this suggestion, but after setting `_BrowserHistory` directly to `createHistory()`, I get a type error in util/URL.ts. So I'm going to leave this as is for now.
<img width="579" alt="Screen Shot 2022-07-13 at 3 54 21 PM" src="https://user-images.githubusercontent.com/6411206/178832623-8029724b-00d5-40ff-8533-8ab49275601f.png">

labkey-alan replied:
Fine to leave it as is for now, but we really should figure out why that is happening. Some type somewhere is wrong, and that needs to be fixed. We may need to update the library we use, or the types library associated with it.

labkey-alan reviewed (changes requested):
Overall this looks great, I did highlight a few issues that I think should be addressed though

cnathe committed:
Merge branch 'develop' into fb_removeReactnDep

# Conflicts:
#	packages/components/package.json
#	packages/components/releaseNotes/components.md

cnathe committed:
v2.198.0-fb-removeReactnDep.0

cnathe committed:
DetailEditRenderer.tsx fix for displaying date picker input for case where initial value is empty (regression from changes in 2.196.1)

cnathe committed:
v2.198.0-fb-removeReactnDep.1

cnathe committed:
v2.198.0-fb-removeReactnDep.2

cnathe committed:
CR feedback:
- rename and move getDeleteErrorNotification and getDeleteSuccessNotification
- remove unneeded _BrowserHistory.listen()
- jest test updates for jest.fn - jest.fn()

cnathe committed:
Refactor message prop for NotificationItemProps to be ReactNode only

cnathe committed:
add back BrowserHistory listen()

cnathe committed:
jest test update

cnathe committed:
Merge branch 'develop' into fb_removeReactnDep

# Conflicts:
#	packages/components/package.json
#	packages/components/releaseNotes/components.md

cnathe committed:
v2.198.1-fb-removeReactnDep.0

cnathe committed:
revert removal of the message callback function for LKB case

cnathe committed:
v2.198.1-fb-removeReactnDep.1

cnathe committed:
pass container to getDateFormat() in Notifications.tsx

cnathe committed:
missed from prev commit

cnathe committed:
Merge branch 'develop' into fb_removeReactnDep

# Conflicts:
#	packages/components/package.json
#	packages/components/releaseNotes/components.md

cnathe committed:
yarn run lint-branch-fix

cnathe committed:
yarn run lint-branch-fix

cnathe committed:
Update releaseNotes with version number and release date

cnathe committed:
v2.200.0
